design-proposals/vhostuser-interface.md

```diff
@@ -0,0 +1,96 @@
+
+## Motivation
+- Any kubevirt VM uses linux kernel interfaces to connect to its host and to reach out externally or to communicate within the cluster. The current kubevirt VMs can only support the kernel data path for which the traffic needs to be through the kernel of the node/compute hosting the VM.
+
+- As a result, when we have a host with DPDK dataplane running in Userspace, the traffic from kubevirt VM will go through the kernel and reach the dataplane running in userspace. This longer path slows down the traffic between the application and the dataplane when both are running in userspace.
+
+- In order to improve the traffic performance and make use of the dpdk, we need to use a different path that can by-pass the kernel and reach the dataplane from the user space directly. This so called fast path will run within the user space and will allow fast packet processing.
+
+## Goals
+- The kubevirt VM when created with DPDK features, should be able to use fast datpath if the host supports the DPDK dataplane allowing fast packet processing to improve performance of the traffic.
+
+## Non Goals
+- The user space networking will require the host to a DPDK dataplane.
+- The kubevirtci creates kernel mode dataplane hosts for testing the VMs. It needs to be updated to support DPDK.
+
+## Definition of Users
+- Any user who has a (k8s) cluster with a dpdk dataplane can use the interface/feature.
+- The feature will be supported only if a valid CNI is being used.
+
+## User Stories
+- As a user/admin, I want to use user space networking feature if I have DPDK supported VM and the host has a DPDK dataplane.
+
+## Repos
+
+- kubevirt/kubevirt
+   - Introduction of vhostuser type interface.
+- kubevirt/kubevirtci
+   - Enable usage of dpdk on the setups created by kubevirtci.
+
+# Design
+- The design involves introducing vHostUser interface type and required parameters to support it. An EmptyDIR volume (shared-dir) and DownwardAPI (podinfo) volume are mounted on the virtlauncher pod to have the support for new interface.
+- These mounts will allow the virt-launcher pod to create a VM with an additional interface which can be reached using the vhostSocket mentioned in the DownwardAPI.
+
+- Creating the VMs will follow the same process with a few changes highlighted as below:
+    1. **Once the VM spec virt-controller will add two new volumes to the virt-launcher pod.**\
+        a. **EmptyDir volume named (shared-dir) is used to share the vhostuser socket file with the virt-launcher pod and dpdk dataplane.**\
+        b. **DownwardAPI volume named (podinfo) is used to share vhostuser socket file name with the virt-launcher via pod annotations.**
+    2. **The CNI will mount the shared-dir i.e., EmptyDir volume with the mount-path as "/var/lib/cni/usrcni/".**
```

Comment:
Is this mount transparent to k8s? If not, we can end up in a clean-up hell. Let's imagine that the CNI plugin is not responding and doesn't perform the unmount, what will happen if we want to delete the virt-launcher pod?

Reply:
Since, k8s deletes the emptydir once the pod is deleted, the deletion of virtualization-launcher pod would not cause any issues.

Reply:
Keep in mind CNI deletes can be missed / fail to cleanup all resources on the pod.

Reply:
I see 2 way for solving this:
 1. Specify the socket path in KubeVirt CR (as a new field) and we let virt-handler mount the socket into virt-launcher. This can follow what we have for container and hotplugged disk
 2. Introduce a new device plugin that mounts the socket. This is at least transparent to k8s and it could avoid the DownwardAPI as you can pass the socket path as an environmental variable.

I used the second approach for the SCSI reservation https://github.com/kubevirt/kubevirt/pull/9177. You could implemented as an external plugin and then we just need to teach virt-launcher to use the new environmental variable

Reply:
Given these 2 choices, I'd go for the latter (device plugin) since it's native to kubernetes.

Reply:
The downside with the device plugin approach is that is implying exposing a resource on the node.  You might or might not expose it in this way. That was one of the reason I was asking about the scheduling [here](https://github.com/kubevirt/community/pull/218#issuecomment-1532602246)

Reply:
@alicefr , @maiqueb,

The main reason I am using EmptyDir and DownwardAPI in my implementation is that they allow transparency between CNI and the kubemanager.

I am not aware of Device Plugin type, will I be able to achieve the transparency between CNI and the kubemanager as they need to use the same name i.e., they should have the same name for the vhostSocket file.